walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,75 @@
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.Collections;
+import java.util.ArrayList;
+import java.util.List;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    List<TimeRange> possibleTimes = new ArrayList<>();
+    List<TimeRange> conflictingTimes = new ArrayList<>();
+    Collection<String> attendees = request.getAttendees();
+
+    // Find every TimeRange that needs to be avoided
+    for (Event event : events) {
+      for (String attendee : event.getAttendees()) {
+        if (attendees.contains(attendee)) {
+          conflictingTimes.add(event.getWhen());
+          break;
+        }
+      }
+    }
+
+    Collections.sort(conflictingTimes, TimeRange.ORDER_BY_START);
+
+    // Remove duplicates and overlaps
+    for (int i = 0; i < conflictingTimes.size() - 1; i++) {
+      TimeRange first = conflictingTimes.get(i);
+      TimeRange second = conflictingTimes.get(i + 1);
+
+      if (first.overlaps(second)) {
+        if (!first.contains(second)) {
+          TimeRange combined = TimeRange.fromStartEnd(first.start(), second.end(), false);
+          conflictingTimes.set(i, combined);
+        }
+
+        conflictingTimes.remove(second);
+        i--;
+      }
+    }
+
+    // Find every TimeRange that doesn't have a conflict
+    if (conflictingTimes.isEmpty()) {
+      possibleTimes.add(TimeRange.WHOLE_DAY);
+    } else {
+      TimeRange initial = conflictingTimes.get(0);
+      TimeRange valid = TimeRange.fromStartEnd(TimeRange.START_OF_DAY, initial.start(), false);
+      possibleTimes.add(valid);
+
+      for (int i = 0; i < conflictingTimes.size() - 1; i++) {
+        TimeRange first = conflictingTimes.get(i);
+        TimeRange second = conflictingTimes.get(i + 1);
+
+        valid = TimeRange.fromStartEnd(first.end(), second.start(), false);
+        possibleTimes.add(valid);
+      }
+
+      TimeRange last = conflictingTimes.get(conflictingTimes.size() - 1);
+      valid = TimeRange.fromStartEnd(last.end(), TimeRange.END_OF_DAY, true);
+      possibleTimes.add(valid);
+    }
+
+    // Remove every TimeRange that is too short
+    long duration = request.getDuration();
+
+    for (int i = 0; i < possibleTimes.size(); i++) {
+      if (possibleTimes.get(i).duration() < duration) {
+        possibleTimes.remove(i);
+        i--;
+      }
+    }
```

Comment:
Maybe we can avoid another loop by checking the duration above before adding the time to possibleTimes?